### PR TITLE
SCP-2151: PAB handles contract logs as part of its own logging

### PR DESCRIPTION
**Pull request by j-mueller**
* Log contract messages in the PAB's logging system (in addition to writing them into the contract state)
* This involved adding a `lastLogs` field with only the new messages, to avoid logging all the messages on every step
* Also had to shuffle some types around in `plutus-pab` to avoid cyclical imports

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: marlowe/pab/Main.hs [hunk 1]
```diff
@@ -51,6 +53,7 @@ instance Pretty Marlowe where
     pretty = viaShow
 handleMarloweContract ::
     ( Member (Error PABError) effs
+    , Member (LogMsg (PABMultiAgentMsg (Builtin Marlowe))) effs
```
Comment (raduom) [inline marlowe/pab/Main.hs:56]: Where is this effect used?

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs [hunk 2]
```diff
@@ -90,7 +90,7 @@ activateContractSTM ::
 activateContractSTM runAppBackend a@ContractActivationArgs{caID, caWallet} = do
     activeContractInstanceId <- ContractInstanceId <$> uuidNextRandom
     logDebug @(ContractInstanceMsg t) $ InitialisingContract caID activeContractInstanceId
-    initialState <- Contract.initialState @t caID
+    initialState <- Contract.initialState @t activeContractInstanceId caID
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs:93]: I suspect the `id` was added here to be able to log it?

File: plutus-pab/src/Plutus/PAB/Monitoring/PABLogMsg.hs [hunk 3]
```diff
@@ -74,10 +77,10 @@ instance Pretty (ContractDef t) => Pretty (AppMsg t) where
         ContractInstances t s            -> pretty t <+> pretty s
         ContractHistoryItem instanceId s -> pretty instanceId <> colon <+> pretty (fmap encodeToLazyText s)
 
-data PABLogMsg =
+data PABLogMsg t =
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Monitoring/PABLogMsg.hs:80]: Why did you add this parameter?

**Review by raduom — APPROVED** [review]
LGTM

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs:93]: yes that's correct

File: plutus-pab/src/Plutus/PAB/Monitoring/PABLogMsg.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Monitoring/PABLogMsg.hs:80]: `t` is used throughout the PAB as the type of contracts that are run by the PAB. It's usually either `ContractExe` (for contracts that exist as executables somewhere in the file system) or `Builtin` (for contracts that are compiled with the PAB and run in the same process)

Previously this was specialised to `ContractExe` in `PABLogMsg` but I needed to use it with other types as well.

File: marlowe/pab/Main.hs (hunk 1 above)
Comment (j-mueller) [reply marlowe/pab/Main.hs:56]: It's used by `Builtin.handleBuiltin`, to log the new messages in `lastLogs`. Before this PR the messages weren't logged with the `LogMsg` effect at all, they just existed somewhere in the contract state.

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs:93]: so when we write the logs to the terminal we can see which instance they came from (useful if there are multiple instances)
